akka-projection-core/src/main/java/akka/projection/internal/spi/SingleShotProjectionMetrics.java

```diff
@@ -0,0 +1,40 @@
+package akka.projection.internal.spi;
+
+import akka.actor.ClassicActorSystemProvider;
+
+public interface SingleShotProjectionMetrics {
+
+    /**
+     *  Must be invoked for each element processed successfully only when the associated offset has been committed.
+     *
+     * @param projectionName the projection name
+     * @param projectionKey  the unique key for this Projection instance
+     * @param waitTime if the Source or the stream element provide the timestamp of the instant when the message
+     *                 was inserted in the queue, the Wait Time is the period between that timestamp and the
+     *                 instant when the event was read from the queue. Set to null if the the insertion
+     *                 timestamp is not available.
+     * @param processTime period of time between the instant when the element was read from the Source until
+     *                    the instant immediately after the offset associated to the element was committed.
+     * @param systemProvider a `ClassicActorSystemProvider` for telemetry to extract/set data on the ActorSystem
+     */
+    void onProcessComplete(String projectionName,
+                           String projectionKey,
+                           Long waitTime,
+                           Long processTime,
+                           ClassicActorSystemProvider systemProvider);
+
+    /**
+     * Must be invoked when the stream fails.
+     *
+     * @param projectionName the projection name
+     * @param projectionKey  the unique key for this Projection instance
+     * @param cause the cause of the failure
+     * @param systemProvider a `ClassicActorSystemProvider` for telemetry to extract/set data on the ActorSystem
+     */
+    void onFailure(String projectionName,
+                   String projectionKey,
+                   Throwable cause,
+                   ClassicActorSystemProvider systemProvider);
+
```

Comment:
This is a simplification of the above in which the responsibility to measure periods falls on the projection implementor. Then, there's a single invocation required in the happy path. Here's an example usage:

```scala
Source
  .futureSource(futSource)
  .mapAsync(1) { elt =>
    val before = now
    processElement(elt).map { elt =>
      metrics.onProcessComplete(projectionId, insertionTimeExtractor(elt), period(before), systemProvider)
    }
  }
```

while this looks appealing it takes away some information from telemetry. Since there's no invocation before `processElement` there's no chance for telemetry to, for example, start an OpenTracing span or any future information cinnamon provides. Instead, just two `Long`s are passed in.

Reply:
I prefer the first approach, not only because it's better for Cinnamon, but also more obvious for implementors.

Reply:
I agree, we should have the separate methods onProcessStart, onProcessComplete and so on